pkg/controller/genericwebhookconfiguration/certhandler.go

```diff
@@ -0,0 +1,324 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package genericwebhookconfiguration
+
+import (
+	"fmt"
+	"strings"
+
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/kubernetes-sigs/kubebuilder/pkg/webhook/certprovisioner"
+	"net/url"
+)
+
+var (
+	// Use an annotation in the following format:
+	// secret.certprovisioner.kubernetes.io/<webhook-name>: <secret-namespace>/<secret-name>
+	// the webhook cert manager library will provision the certificate for the webhook by
+	// storing it in the specified secret.
+	SecretCertInjectionAnnotationKeyPrefix = "secret.certprovisioner.kubernetes.io/"
+
+	// Use an annotation in the following format:
+	// local.certprovisioner.kubernetes.io/<webhook-name>: path/to/certs/
+	// the webhook cert manager library will provision the certificate for the webhook by
+	// storing it under the specified path.
+	// format: local.certprovisioner.kubernetes.io/webhookName: path/to/certs/
+	LocalCertInjectionAnnotationKeyPrefix = "local.certprovisioner.kubernetes.io/"
+
+	CACertName     = "ca-cert.pem"
+	ServerKeyName  = "key.pem"
+	ServerCertName = "cert.pem"
+)
```

Comment:
I tried understanding the high level functionality it provides, but couldn't :) would help if we can have a blurb somewhere explaining what are the components involved.

Reply:
So far what I have got is that.... this is sort of a live component running in the cluster provisioning SSL stuff required by a webhook. Webhook will signal the secret & path details using the annotation (which resource though not clear ?)

Reply:
There are at least 2 typical ways to consume this lib:
- Run it as a controller in an container (either in a standalone pod or sidecar with the webhook server). `pkg/controller/mutatingwebhookconfiguration/controller.go` is an example.
- Invoke it as a lib. e.g. Build your `GenericWebhookConfigurationController` and call `Sync(webhookConfiguration runtime.Object)` method.